File: solutions/_983_minimum_cost_for_tickets/solution.py
from typing import List


class Solution:
    def mincostTickets(self, days: List[int], costs: List[int]) -> int:
        dp = [0] * (days[-1] + 1)
        i = 0
        for day in range(1, len(dp)):
            if day < days[i]:
                dp[day] = dp[day - 1]
            else:
                i += 1
                dp[day] = min(
                    dp[day - 1] + costs[0],
                    dp[max(day - 7, 0)] + costs[1],
                    dp[max(day - 30, 0)] + costs[2],
                )
        return dp[-1]


# https://leetcode.com/problems/minimum-cost-for-tickets/

File: solutions/_983_minimum_cost_for_tickets/solution.js
/**
 * @param {number[]} days
 * @param {number[]} costs
 * @return {number}
 */
var mincostTickets = function (days, costs) {
  const dp = new Array(days.at(-1) + 1).fill(0);
  const set = new Set(days);
  for (let i = 1; i < dp.length; i++) {
    if (!set.has(i)) {
      dp[i] = dp[i - 1];
    } else {
      dp[i] = Math.min(
        dp[i - 1] + costs[0],
        dp[Math.max(0, i - 7)] + costs[1],
        dp[Math.max(0, i - 30)] + costs[2]
      );
    }
  }
  return dp.at(-1);
};

module.exports = { mincostTickets };

/*
https://leetcode.com/problems/minimum-cost-for-tickets/
*/
